crates/cadenza-syntax/src/parse.rs

```diff
@@ -116,7 +116,62 @@ impl<'src> Parser<'src> {
 
         // Loop to handle sequences of operators and function application
         loop {
-            // Check what comes next
+            // Special case: Check for array indexing (arr[0]) before skipping trivia
+            // Array indexing requires no whitespace before '[', otherwise it's function application
+            if self.current() == Kind::LBracket {
+                // Check if we should continue parsing at all (handles indentation/line breaks)
+                // This is important: even if we see '[', we need to respect marker boundaries
+                if !marker.should_continue(self) {
+                    marker.finish(self);
+                    return;
+                }
+
+                // Check if there's whitespace immediately before '['
+                // The whitespace tracker contains the span of the most recent whitespace
+                // If the span is non-empty, there was whitespace consumed
+                let has_whitespace_before = self.whitespace.span.start != self.whitespace.span.end;
+
+                // Only treat as indexing if there's no whitespace before '['
+                if !has_whitespace_before {
+                    // Array indexing binding power (same as PathAccess)
+                    let l_bp = 34;
```

Comment:
We really shouldn't hard code binding powers here, especially when they could change. Instead we should update the token codegen to have a way to get the value. Maybe we have a new enum with all of the categories and then it's not bound to any particular token